sklearn/linear_model/tests/test_ridge.py

```diff
@@ -311,6 +312,108 @@ def test_ridge_individual_penalties():
+    )
+    X += 30 * np.random.RandomState(0).randn(X.shape[1])
+
+    alphas = [1e-3, .1, 1., 10., 1e3]
+    loo_ridge = RidgeCV(cv=n_samples, fit_intercept=fit_intercept,
+                        alphas=alphas, scoring='neg_mean_squared_error',
+                        normalize=normalize)
+    gcv_ridge = RidgeCV(gcv_mode=gcv_mode, fit_intercept=fit_intercept,
+                        alphas=alphas, normalize=normalize)
+
+    loo_ridge.fit(X, y)
+
+    X_gcv = X_constructor(X)
+    gcv_ridge.fit(X_gcv, y)
+
+    assert gcv_ridge.alpha_ == pytest.approx(loo_ridge.alpha_)
+    assert_allclose(gcv_ridge.coef_, loo_ridge.coef_, rtol=1e-4)
+    assert_allclose(gcv_ridge.intercept_, loo_ridge.intercept_, rtol=1e-4)
+
+
+@pytest.mark.parametrize('gcv_mode', ['svd', 'eigen'])
+@pytest.mark.parametrize('X_constructor', [np.asarray, sp.csr_matrix])
+@pytest.mark.parametrize('fit_intercept', [True, False])
+@pytest.mark.parametrize('n_features', [11, 69])
+@pytest.mark.parametrize('noise', [1., 30.])
+def test_ridge_gcv_sample_weights(
+        gcv_mode, X_constructor, fit_intercept, n_features, noise):
+    alphas = [1e-3, .1, 1., 10., 1e3]
+    rng = np.random.RandomState(0)
+    x, y = datasets.make_regression(
+        n_samples=59, n_features=n_features, n_targets=4,
+        random_state=0, shuffle=False, noise=noise)
+    x += 30 * rng.randn(x.shape[1])
+    sample_weight = 3 * rng.randn(len(x))
+    sample_weight = (sample_weight - sample_weight.min() + 1).astype(int)
+    indices = np.repeat(np.arange(x.shape[0]), sample_weight)
+    sample_weight = sample_weight.astype(float)
+    X_tiled, y_tiled = x[indices], y[indices]
+
+    cv = GroupKFold(n_splits=x.shape[0])
+    splits = cv.split(X_tiled, y_tiled, groups=indices)
+    kfold = RidgeCV(
+        alphas=alphas, cv=splits, scoring='neg_mean_squared_error',
+        fit_intercept=fit_intercept)
+    with ignore_warnings():
+        kfold.fit(X_tiled, y_tiled)
+
+    ridge_reg = Ridge(alpha=kfold.alpha_, fit_intercept=fit_intercept)
+    splits = cv.split(X_tiled, y_tiled, groups=indices)
+    predictions = cross_val_predict(ridge_reg, X_tiled, y_tiled, cv=splits)
+    kfold_errors = (y_tiled - predictions)**2
+    kfold_errors = [
+        np.sum(kfold_errors[indices == i], axis=0) for
+        i in np.arange(x.shape[0])]
+    kfold_errors = np.asarray(kfold_errors)
+
+    x_gcv = X_constructor(x)
+    gcv_ridge = RidgeCV(
+        alphas=alphas, store_cv_values=True,
+        gcv_mode=gcv_mode, fit_intercept=fit_intercept)
+    gcv_ridge.fit(x_gcv, y, sample_weight=sample_weight)
+    gcv_errors = gcv_ridge.cv_values_[:, :, alphas.index(kfold.alpha_)]
+
+    assert kfold.alpha_ == gcv_ridge.alpha_
+    assert_allclose(gcv_errors, kfold_errors, rtol=5e-2)
+    assert_allclose(gcv_errors, kfold_errors, rtol=5e-2)
+    assert_allclose(gcv_ridge.coef_, kfold.coef_, rtol=5e-2)
+    assert_allclose(gcv_ridge.intercept_, kfold.intercept_, rtol=5e-2)
+
+
+def test_check_gcv_mode():
```

Comment:
you can change the test as:

```python
@pytest.mark.parametrize('mode', [True, 1, 5, 'bad', 'gcv', np.arange(3)])
def test_check_gcv_mode_error(mode):
    X, y = make_regression(n_samples=5, n_features=2)
    gcv = RidgeCV(gcv_mode=mode)
    with pytest.raises(ValueError, match="Unknown value for 'gcv_mode'"):
        gcv.fit(X, y)
        _check_gcv_mode(X, mode)


@pytest.mark.parametrize(
    'mode, mode_samples_sup_features, mode_features_sup_samples',
    [(None, 'svd', 'eigen'),
     ('auto', 'svd', 'eigen'),
     ('eigen', 'eigen', 'eigen'),
     ('svd', 'svd', 'svd')]
)
def test_check_gcv_mode_choice(mode, mode_samples_sup_features,
                               mode_features_sup_samples):
    X, _ = make_regression(n_samples=5, n_features=2)

    assert _check_gcv_mode(X, mode) == mode_samples_sup_features
    assert _check_gcv_mode(X.T, mode) == mode_features_sup_samples
```

Reply:
thanks!


> ```python
>     with pytest.raises(ValueError, match="Unknown value for 'gcv_mode'"):
>         gcv.fit(X, y)
>         _check_gcv_mode(X, mode)
> ```

I never used this context manager before; both lines here need to raise (ie `gcv.fit` and `_check_gcv_mode` should both fail). does the second line get executed if the first one fails?

Reply:
I supposed that yes but let me check to be sure. Otherwise we call twice the context manager.

Reply:
So you are right, we need to split the 2 lines

```python
@pytest.mark.parametrize('mode', [True, 1, 5, 'bad', 'gcv', np.arange(3)])
def test_check_gcv_mode_error(mode):
    X, y = make_regression(n_samples=5, n_features=2)
    gcv = RidgeCV(gcv_mode=mode)

    err_msg = "Unknown value for 'gcv_mode'"
    with pytest.raises(ValueError, match=err_msg):
        gcv.fit(X, y)
    with pytest.raises(ValueError, match=err_msg):
        _check_gcv_mode(X, mode)
```

Reply:
No the second line will not be called if the first raises an exception. To check both, you need to use the context manager twice.